polytracker/include/polytracker/passes/tainted_control_flow.h

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2023-present, Trail of Bits, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed in accordance with the terms specified in
+ * the LICENSE file found in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <llvm/IR/InstVisitor.h>
+#include <llvm/IR/PassManager.h>
+#include <unordered_map>
+
+namespace polytracker {
+namespace detail {
+struct FunctionMappingJSONWriter;
+}
+
+class TaintedControlFlowPass
+    : public llvm::PassInfoMixin<TaintedControlFlowPass>,
+      public llvm::InstVisitor<TaintedControlFlowPass> {
+  //
+  llvm::IntegerType *label_ty{nullptr};
+  // Taint tracking startup
+  llvm::FunctionCallee taint_start_fn;
+  // Log taint label affecting control flow
+  llvm::FunctionCallee cond_br_log_fn;
+  // Log enter/leave functions
+  llvm::FunctionCallee fn_enter_log_fn;
+  llvm::FunctionCallee fn_leave_log_fn;
+
+  // Helpers
+  void insertCondBrLogCall(llvm::Instruction &inst, llvm::Value *val);
+  void insertTaintStartupCall(llvm::Module &mod);
+  void declareLoggingFunctions(llvm::Module &mod);
+
+  llvm::ConstantInt *get_function_id_const(llvm::Function &f);
+  llvm::ConstantInt *get_function_id_const(llvm::Instruction &i);
+
+public:
+  using function_id = uint32_t;
+
+  TaintedControlFlowPass();
```

Comment:
Me too, but I think in this case we need to have a ctor as it does some initialization (see cpp-file). Otherwise we would have to refactor the code to initialize on first use or something like that. I don't think thats a terribly worthwhile investment, but I could be wrong.

Reply:
I see! Hmm, could we move the initialization of the JSON writer to the `run()` method? That way we can get rid of the default declarations.